cpp/src/parquet/page_index.h

```diff
@@ -0,0 +1,143 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "parquet/platform.h"
+#include "parquet/statistics.h"
+#include "parquet/types.h"
+
+namespace parquet {
+
+/// \brief BoundaryOrder is a proxy around format::BoundaryOrder.
+enum class PARQUET_EXPORT BoundaryOrder { UNORDERED = 0, ASCENDING = 1, DESCENDING = 2 };
+
+/// \brief ColumnIndex is a proxy around format::ColumnIndex.
+class PARQUET_EXPORT ColumnIndex {
+ public:
+  /// \brief Create a ColumnIndex from a serialized thrift message.
+  static std::unique_ptr<ColumnIndex> Make(const ColumnDescriptor* descr,
+                                           const void* serialized_index,
+                                           uint32_t* inout_index_len,
+                                           const ReaderProperties& properties);
+
+  virtual ~ColumnIndex() = default;
+
+  /// \brief Returns number of pages in this column index.
+  virtual int64_t num_pages() const = 0;
+
+  /// \brief Returns if all values are null in a single page.
+  virtual bool null_page(int64_t page_id) const = 0;
+
+  /// \brief Returns whether both min_values and max_values are
+  /// orderd and if so, in which direction.
+  virtual BoundaryOrder boundary_order() const = 0;
+
+  /// \brief Returns if null count is available.
+  virtual bool HasNullCount() const = 0;
+
+  /// \brief Returns null count for a single page.
+  virtual int64_t null_count(int64_t page_id) const = 0;
+
+  /// \brief Returns all null indicator for each page in batch.
+  virtual const std::vector<bool>& GetNullPages() const = 0;
+
+  /// \brief Returns null count for each page in batch.
+  virtual const std::vector<int64_t>& GetNullCounts() const = 0;
+
+  /// \brief The minimum value of a single page. Throws if it is null page.
+  virtual std::string GetEncodedMin(int64_t page_id) const = 0;
+
+  /// \brief The maximum value of a single page. Throws if it is null page.
+  virtual std::string GetEncodedMax(int64_t page_id) const = 0;
+};
+
+/// \brief Typed implementation of ColumnIndex.
+template <typename DType>
+class PARQUET_EXPORT TypedColumnIndex : public ColumnIndex {
+ public:
+  using T = typename DType::c_type;
+
+  /// \brief The minimum value of a single page. Throws if it is null page.
+  virtual T min_value(int64_t page_id) const = 0;
+
+  /// \brief The maximum value of a single page. Throws if it is null page.
+  virtual T max_value(int64_t page_id) const = 0;
+
+  /// \brief The minimum value of every valid page.
+  virtual const std::vector<T>& GetMinValues() const = 0;
+
+  /// \brief The maximum value of every valid page.
+  virtual const std::vector<T>& GetMaxValues() const = 0;
+
+  /// \brief Returns list of page index of all valid pages.
```

Comment:
Is it useful to have this in addition from `null_pages`?

Reply:
This makes it easy to know which page_id is valid when the return values from line 92 are used.